### PR TITLE
New random module to wrap stream module

### PR DESCRIPTION
**Pull request by 1cosmologist**
This PR adds new higher level random module for overall random stream management. As a part of this effort we checked complex and real random number reproducibility across different hardware. This detailed in issue https://github.com/exgalsky/xgmutil/issues/1 . Based on this, by deafult all random number is generated at 64bits.

1. The `random` module now manages the random number streams based on single initial seed with accessible metadata.
2. The `RNG_manager` class creates and manages all `RNG_component` classes for each individual stream.
3. The `stream` module still handles the scalable random number generation as before but `stream.Stream` does to take a seed directly instead using a `PRNGKey`, that is passed from `random` module.  The `stream.Stream` allows setting the jax numpy data type of the random number stream.
4. `Stream.generate` uses new kwarg `mc` to specify realization of the random stream. If the start and size remain same, `mc` allows deterministic generation of new random numbers.

Tested on CPU and GPU.

### CODE REVIEW

File: xgmutil/stream.py [hunk 1]
```diff
@@ -1,26 +1,33 @@
...
 import jax
-import sys
 import os
 
 class Stream:
     '''Stream'''
     def __init__(self, **kwargs):
 
         self.force_no_gpu = kwargs.get('force_no_gpu',False)
-        self.seed         = kwargs.get('seed',123456789)
```
Comment (marcelo-alvarez) [inline xgmutil/stream.py:12]: @1cosmologist can we keep this variable called seed in the initialization and allow for it to be either an integer (to retain the functionality before this change) or a key (as you have it now)?

File: xgmutil/stream.py [hunk 2]
```diff
@@ -29,9 +36,7 @@ def generate(self,**kwargs):
 
         seqIDs = jnp.arange(start_seqID, end_seqID+1)
 
-        seedkey = rnd.PRNGKey(self.seed)
```
Comment (marcelo-alvarez) [inline xgmutil/stream.py:32]: here we should change the behavior depending on the type of `self.seed`, see my comment above.

File: xgmutil/stream.py [hunk 3]
```diff
@@ -1,26 +1,33 @@
...
         if self.force_no_gpu:
-            _JAX_PLATFORM_NAME = os.environ["JAX_PLATFORM_NAME"]
-            os.environ["JAX_PLATFORM_NAME"] = "cpu"
+            _JAX_PLATFORM_NAME = jax.default_backend()
+            jax.default_device("cpu")
 
         start = kwargs.get('start',0)
         size  = kwargs.get('size' ,1)
         dist  = kwargs.get('dist','normal')
+        mc    = kwargs.get('mc', 0)
```
Comment (marcelo-alvarez) [inline xgmutil/stream.py]: If I understand correctly, to retain the previous functionality as an option (i.e. user provides an integer seed instead of a key), then the `mc` variable can assumed to be zero in the case where `self.seed` is an integer?

**Review by marcelo-alvarez — CHANGES_REQUESTED** [review]
Thanks for this @1cosmologist, it looks like a nice improvement to add more generality to the existing functionality. Before testing the downstream effects of these changes in the overall sky modeling pipeline, I would like to be able to retain the previous behavior of the `Stream` class, i.e. to provide an integer seed, for maximum flexibility in the future. Please see my comments inline.

File: xgmutil/stream.py (hunk 3 above)
Comment (1cosmologist) [reply xgmutil/stream.py]: I presume if we have a `seed` key, the `mc` key can be degenerate depending on how it is used. One can generate new realizations by changing the seed (or the PRNGKey) itself and creating a new stream. But I would advice against that. The intended used would be to initialize once with a seed or PRNGKey and then independent realizations by changing `mc`. I think by default `mc` is always assumed to be zero, whether we use PRNGKey or seed. I hope that answers your question.

File: xgmutil/stream.py (hunk 1 above)
Comment (1cosmologist) [reply xgmutil/stream.py:12]: Ok, it can be added. But which takes precedence when both are provided?

File: xgmutil/stream.py (hunk 2 above)
Comment (1cosmologist) [reply xgmutil/stream.py:32]: ok, but we need to set a precedence in case both are provided.

**Comment by 1cosmologist** [comment]
Updated stream to retain old functionality (https://github.com/exgalsky/xgmutil/pull/2/commits/7d6642fbda660216e1c95c265c17068c2a64e6d9). Stream can now take an integer seed or a PRNGKey. It also has a method to update the key with fold_in of an integer. The only API change affecting old use of stream is that the `seed` kwarg has been replaced by `seedkey`.

The higher level API through the random.py is unchanged. Though internally the function of updating the PRNGKey is called from the `generate` method.

File: xgmutil/stream.py (hunk 2 above)
Comment (marcelo-alvarez) [reply xgmutil/stream.py:32]: looks done, except `self.seed` --> `self.seedkey`, resolving...

File: xgmutil/stream.py (hunk 1 above)
Comment (marcelo-alvarez) [reply xgmutil/stream.py:12]: This has been resolved by using the type of `self.seedkey` to determine how it works.

File: xgmutil/stream.py (hunk 3 above)
Comment (marcelo-alvarez) [reply xgmutil/stream.py]: Thanks for the explanation. I now understand that `mc` is an integer that uniquely labels a given realization, which is folded into a given PRNGKey. As is, if `self.seedkey` is an integer, then the value of `mc` will have no effect, preserving the existing functionality in main that allows the more straightforward approach of providing an integer seed without having to pass in a key.

**Review by marcelo-alvarez — APPROVED** [review]

**Comment by marcelo-alvarez** [comment]
Thanks @1cosmologist for the changes. They look good, but I will first verify that the fiducial kappa map pipeline still runs as expected when changing `seed` to `seedkey` when initializing a `Stream` object.

**Comment by 1cosmologist** [comment]
It should run but the map would be different from before as we generate float32 random number as float64 now. This would mean different random numbers from before.

**Comment by marcelo-alvarez** [comment]
> It should run but the map would be different from before as we generate float32 random number as float64 now. This would mean different random numbers from before.

Is there an option to change to change to float32 and make it the default (since our current default use case -- massively parallel 3D gaussian random fields for LSS sims -- uses float32 due to memory constraints)?

**Comment by 1cosmologist** [comment]
The output can be float32 instead of float64, for later computation requirements. But the numbers are always generated as float64. The current implementation does not allow for a switch to change this default behaviour. This ensures that the if we want to do the simulation at higher precision, the numbers don't change just have more digits.

**Comment by marcelo-alvarez** [comment]
Understood, thanks. Tested with 
```
srun -n 4 mockgen mockgen-test
```
after modifying lpt main branch to use `xgmutil` instead of `scaleran` and the results looks fine to me. Merging now.
